drivers/soundwire/master.c

```diff
@@ -14,9 +15,15 @@ static void sdw_master_device_release(struct device *dev)
 	kfree(md);
 }
 
+static const struct dev_pm_ops master_dev_pm = {
+	SET_RUNTIME_PM_OPS(pm_generic_runtime_suspend,
+			   pm_generic_runtime_resume, NULL)
+};
+
 struct device_type sdw_master_type = {
 	.name =		"soundwire_master",
 	.release =	sdw_master_device_release,
+	.pm = &master_dev_pm,
```

Comment:
Interesting, I'm curious how this works. `pm_generic_runtime_*()` do just one thing - they call the respective device runtime PM method. These PM methods are now assigned to the device type. So when `__rpm_get_callback()` selects a PM method to call, these device type methods will override bus type and device class rt-pm ops. Is this the goal of this change? What happens without it - which runtime PM operation gets called? `sdw_bus_type` doesn't have PM ops either, so, without this patch device own method should be called anyway?

Reply:
@lyakh This is to enable PM support for sdw master device. The issue is that we have runtime pm callback functions on both bus and slave driver. And the parent/child relationship is
```
bus device
    -> master device
        -> slave device
```
We need to wake up the bus when a slave interrupt is triggered. We will call [pm_request_resume(&slave->dev)](https://github.com/thesofproject/linux/blob/topic/sof-dev/drivers/soundwire/intel.c#L1656), but bus device will not be woke up if master device doesn't support runtime pm. Hope this answer your question. :)

Reply:
> nteresting, I'm curious how this works. `pm_generic_runtime_*()` do just one thing - they call the respective device runtime PM method.

@lyakh the problem is that the master device is not really bound to a driver. But since it is the parent of the slave device, its runtime PM has to be enabled in order for the bus device to be resume when the slave is resumed. And we dont really care about doing anything in the PM callbacks for the master device so the generic op would do precisely that ie return 0.

Reply:
@ranj063 ah, so there are no runtime PM callbacks otherwise for that device? So these `pm_generic_runtime_*()` functions aren't really calling driver callbacks, but just returning 0 instead? And without them `RPM_GET_CALLBACK()` would return `NULL` and `rpm_callback()` would then return `-ENOSYS` and suspend would fail? So in fact you don't even need those generic functions, you need trivial `return 0;` callbacks. But well, this does the job too. A comment could be nice though.